### PR TITLE
Allow asking for any variant on an ab_test

### PR DESCRIPTION
**Pull request by redross**
If ab config changes variant configuration for an ab test,
ab lib used to throw errors if `ab_test.some_variant?` code
was invoked. Now it returns false.

Java/Objective-C implementations seem to behave in similar way.

cc @mmozuras

### CODE REVIEW

**Comment by astrauka** [comment]
I would vote to returning false on variant missing and report error in graylog, that would not crash anything and allow us to see that error occured

**Comment by redross** [comment]
The thing is, ab is kind of separate from the actual core codebase, we'd need to add configurable logger to this then. (That does sound like a plan though :+1:  )

**Comment by tomasv** [comment]
@redross ping

**Comment by redross** [comment]
Added to  my to do list.

**Comment by redross** [comment]
@mmozuras @astrauka does this make sense? https://github.com/vinted/ab/commit/3d89e3ae3ad88258b21652fb044e21976a225e69

(will add tests if we agree that's we actually want logging)

**Comment by redross** [comment]
Added some specs, bumped version. Good to merge?

File: spec/assigned_test_spec.rb [hunk 1]
```diff
@@ -16,6 +16,25 @@ module Ab
     let(:buckets) { 1..1000 }
     let(:thousand_variants) { 1.upto(1000).map { |i| AssignedTest.new(test, i).variant } }
 
+    context 'with non existent variant' do
+      let(:variants) { [OpenStruct.new(name: 'enabled', accumulated_chance_weight: 2)] }
+
+      subject { assigned_test.disabled? }
+      it { should be false }
```
Comment (astrauka) [inline spec/assigned_test_spec.rb]: `it { is_expected.to be false }`

**Comment by astrauka** [comment]
Good to me

**Comment by mmozuras** [comment]
@redross tests are failing, try rebasing on `master`.

File: lib/ab/missing_variant.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    private
+
+    def variant_method?(meth)
+      meth.to_s.end_with?('?')
+    end
+
+    def log_missing_variant(meth)
+      return unless self.class.respond_to?(:logger) && self.class.logger
+
+      self.class.logger.debug("[AB_testing] Checking non-existing variant: #{meth}")
```
Comment (mmozuras) [inline lib/ab/missing_variant.rb]: @redross is this sort of logging categorisation (`[AB_testing]`) matches how other gems handle logging?

Comment (tomasv) [inline lib/ab/missing_variant.rb]: I think logging is usually done to a simple `Logger.new(STDOUT)` by default, or a simple stub.

**Comment by redross** [comment]
Will update ;)

File: spec/assigned_test_spec.rb (hunk 1 above)
Comment (redross) [inline spec/assigned_test_spec.rb]: Currently the gem uses should syntax:
```
RSpec.configure do |config|
  config.expect_with(:rspec) { |c| c.syntax = :should }
  config.mock_with(:rspec) { |c| c.syntax = :should }
end
```

Could be changed with separate PR though

**Comment by redross** [comment]
Better? :)

File: lib/ab/tests.rb [hunk 3]
```diff
@@ -8,6 +8,14 @@ def before_picking_variant(&block)
       def after_picking_variant(&block)
         AssignedTest.after_picking_variant(&block)
       end
+
+      def logger=(logger)
+        @logger = logger
+      end
+
+      def logger
+        @logger ||= Logger.new($stdout)
```
Comment (tomasv) [inline lib/ab/tests.rb]: After this we will need to set logger in `core`, right?

File: lib/ab/tests.rb [hunk 4]
```diff
@@ -8,6 +8,14 @@ def before_picking_variant(&block)
       def after_picking_variant(&block)
         AssignedTest.after_picking_variant(&block)
       end
+
+      def logger=(logger)
```
Comment (astrauka) [inline lib/ab/tests.rb]: `attr_writer :logger`

**Comment by astrauka** [comment]
:+1:

File: lib/ab/tests.rb (hunk 3 above)
Comment (redross) [inline lib/ab/tests.rb]: That would make sense, yes.

File: lib/ab/tests.rb [hunk 5]
```diff
@@ -8,6 +8,14 @@ def before_picking_variant(&block)
...
+        @logger = logger
+      end
+
+      def logger
```
Comment (tomasv) [inline lib/ab/tests.rb]: Actually, maybe the logger should be on `Ab` instead of `Ab::Tests`. It looks weird when `Ab::MissingVariant` tries to look in `Ab::Tests` for a logger. We could also use the configuration block pattern for this gem:
```ruby
Ab.configure do |config|
  config.logger = MyLogger.new
end
```

:smile:

File: lib/ab/tests.rb (hunk 3 above)
Comment (mmozuras) [inline lib/ab/tests.rb]: @redross imho, it should default to null instead of `$stdout`. Just like [event-tracker](https://github.com/vinted/event-tracker).

**Comment by redross** [comment]
ping @tomasv

**Comment by tomasv** [comment]
:+1:

File: lib/ab/configuration.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      @config ||= Config.new
+      yield(@config)
+    end
+  end
+
+  class Config
+    attr_accessor :logger
+
+    def initialize
+      @logger = Logger.new(nil)
```
Comment (mmozuras) [inline lib/ab/configuration.rb]: @saulius how is it different than `NullLogger`? Which is better?

Comment (redross) [inline lib/ab/configuration.rb]: Why implement a new logger if the default one behaves like one?

Comment (saulius) [inline lib/ab/configuration.rb]: There are multiple aspects at play here. 

`<oop-zealot-mode-on>`

For one thing being able to pass `nil` to constructor instead of [`LogDevice`](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/lib/logger.rb#L568) is implicit knowledge. If you [take](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/lib/logger.rb#L317-L320) [a careful](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/lib/logger.rb#L366-L368) [look](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/lib/logger.rb#L389-L391) into how that works you'll see that `Logger` itself could use a `NullLogDevice` kind of thing since [its implementation/methods are not sure of themselves](https://github.com/avdi/naught#whats-that-supposed-to-mean). But that's not our problem, back to our problem.

So [if we are really serious about performance](https://github.com/vinted/event-tracker/pull/1#discussion_r29657154) then noop is clearly better than [some](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/lib/logger.rb#L433-L435) [op](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/lib/logger.rb#L364-L368). Also, we [were taught](http://blog.8thlight.com/eric-smith/2011/10/27/thats-not-yours.html) [not to mock](https://github.com/vinted/ab/pull/23/files#diff-cbce597f471835b83022daa27e639b37R23) what we don't own, but in `Loggers` case it's sort of subjective I guess :)

`</oop-zealot-mode-off>`

Just my 2 cents

Comment (tomasv) [inline lib/ab/configuration.rb]: @saulius oh wow, so you wrote a wrapper that does nothing to save an extra stack frame? :smile:

Comment (mmozuras) [inline lib/ab/configuration.rb]: @redross @saulius could we agree on a single way on doing this? :smile:

Comment (saulius) [inline lib/ab/configuration.rb]: To be honest when I was writing it, I just looked at what Avdi did here https://github.com/avdi/naught#whats-that-supposed-to-mean :)

Comment (saulius) [inline lib/ab/configuration.rb]: I don't mind switching to implicit version, after all I'm programming Scala sometimes :trollface:

File: README.md [hunk 7]
```diff
@@ -20,6 +20,11 @@ ab = Ab::Tests.new(configuration, identifier)
 Ab::Tests.before_picking_variant { |test| puts "picking variant for #{test}" }
 Ab::Tests.after_picking_variant { |test, variant| puts "#{variant_name}" }
 
+# by default messages are logged to null logger. Can be changed by setting Your own logger:
```
Comment (mmozuras) [inline README.md]: Why is `Your` capitalised? :smile:

File: lib/ab/version.rb [hunk 8]
```diff
@@ -1,3 +1,3 @@
 module Ab
-  VERSION = '0.2.4'
+  VERSION = '0.2.5'
```
Comment (mmozuras) [inline lib/ab/version.rb]: We bump versions in a separate commit:
https://github.com/vinted/ab/commits/master

**Comment by mmozuras** [comment]
:+1:, except for those two small things.

File: README.md (hunk 7 above)
Comment (redross) [inline README.md]: http://english.stackexchange.com/questions/30185/you-versus-you-as-polite-form-of-writing apparently I mistaken it for a "polite" form.

**Comment by mmozuras** [comment]
:+1:

**Comment by astrauka** [comment]
Finally :)

**Comment by tomasv** [comment]
Just 215 days. :smile:

**Comment by redross** [comment]
Some things take time

Some things take time

**Comment by mmozuras** [comment]
@redross :laughing: :clap:

**Comment by redross** [comment]
Travis seems to be taking forever too..

**Comment by mmozuras** [comment]
@redross one build doesn't seem to start. I'll restart it.

**Comment by mmozuras** [comment]
Pushed the newest version to rubygems.org.

**Comment by redross** [comment]
:bow: :+1:
